nucliadb/nucliadb/common/cluster/manager.py

```diff
@@ -480,45 +480,72 @@ async def add_resource(
             )
 
 
-def choose_node(
+def get_all_shard_nodes(
     shard: writer_pb2.ShardObject,
-    target_replicas: Optional[list[str]] = None,
-    read_only: bool = False,
-) -> tuple[AbstractIndexNode, str, str]:
-    """
-    Choose an arbitrary node storing `shard`.
+    use_read_replicas: bool,
+) -> list[tuple[AbstractIndexNode, str]]:
+    """Return a list of all nodes containing `shard` with the shard replica id.
+    If `use_read_replicas`, read replica nodes will be returned too.
+
     """
-    preferred_nodes = []
-    backend_nodes = []
+    nodes = []
     for shardreplica in shard.replicas:
         node_id = shardreplica.node
         replica_id = shardreplica.shard.id
 
         node = get_index_node(node_id)
         if node is not None:
-            if target_replicas and replica_id in target_replicas:
-                preferred_nodes.append((replica_id, node))
-            else:
-                backend_nodes.append((replica_id, node))
+            nodes.append((node, replica_id))
 
-        if read_only:
+        if use_read_replicas:
             for read_replica_node_id in get_read_replica_node_ids(node_id):
                 read_replica_node = get_index_node(read_replica_node_id)
-                if read_replica_node is None:
-                    continue
-                if target_replicas and replica_id in target_replicas:
-                    preferred_nodes.append((replica_id, read_replica_node))
-                else:
-                    backend_nodes.append((replica_id, read_replica_node))
+                if read_replica_node is not None:
+                    nodes.append((read_replica_node, replica_id))
+
+    return nodes
+
 
-    if len(preferred_nodes) == 0 and len(backend_nodes) == 0:
+def choose_node(
+    shard: writer_pb2.ShardObject,
+    target_replicas: Optional[list[str]] = None,
+    read_only: bool = False,
+) -> tuple[AbstractIndexNode, str, str]:
+    """Choose an arbitrary node storing `shard` following these rules:
+    - nodes containing a shard replica from `target_replicas` are the preferred
+    - if `read_only`, read_replicas are preferred over primaries
+    - if there's more than one option with the same score, a random choice will
+      be made
+
+    According to these rules and considering `read_only == True`, a secondary
+    node containing a shard replica from `target_replicas` is the most
+    preferent, while a primary node with a shard not in `target_replicas` is the
+    least preferent.
+
+    """
+    target_replicas = target_replicas or []
+
+    shard_nodes = get_all_shard_nodes(shard, use_read_replicas=read_only)
```

Comment:
Can we tweak the naming a bit here so that we don't get confused by shard replicas Vs nore read replicas?

Proposal: 
- use `use_secondaries`
- use `shard_replica_id` instead of `replica_id`

Reply:
I used *read replica* because I think it's the most used term but as you may have noticed, we are using *read replica* and *secondaries* in multiple places. IMO we should define which one do we prefer and be consistent. I don't mind which one.

I can change all mentions to `replica` for `shard_replica` though